### PR TITLE
8286122: [macos]: App bundle cannot upload to Mac App Store due to info.plist embedded in java exe

### PR DESCRIPTION
**Pull request by sashamatveev**
- It is not possible to support native JDK commands such as "java" inside Mac App Store bundles due to embedded info.plist. Workarounds suggested in JDK-8286122 does not seems to be visible.
 - With proposed fix we will enforce "--strip-native-commands" option for jlink, so native JDK commands are not included when generating Mac App Store bundles.
 - Custom runtime provided via --runtime-image should not contain native commands as well, otherwise jpackage will throw error.
 - Added two tests to validate fix.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8286122](https://bugs.openjdk.java.net/browse/JDK-8286122): [macos]: App bundle cannot upload to Mac App Store due to info.plist embedded in java exe

### Reviewers
 * [Alexey Semenyuk](https://openjdk.java.net/census#asemenyuk) (@alexeysemenyukoracle - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - Author)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8666/head:pull/8666` \
`$ git checkout pull/8666`

Update a local copy of the PR: \
`$ git checkout pull/8666` \
`$ git pull https://git.openjdk.java.net/jdk pull/8666/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8666`

View PR using the GUI difftool: \
`$ git pr show -t 8666`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8666.diff">https://git.openjdk.java.net/jdk/pull/8666.diff</a>

</details>

### CODE REVIEW

File: test/jdk/tools/jpackage/macosx/MacAppStoreRuntimeTest.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        cmd.executeAndAssertHelloAppImageCreated();
+    }
+
+    @Test
+    public static void test2() throws Exception {
+        JPackageCommand cmd = JPackageCommand.helloAppImage();
+        cmd.addArguments("--mac-app-store", "--runtime-image", getRuntimeImage(false));
+
+        cmd.execute(1);
+    }
```
Comment (alexeysemenyukoracle) [inline test/jdk/tools/jpackage/macosx/MacAppStoreRuntimeTest.java:102]:
```
@Test
@Parameter("true")
@Parameter("false")
public static void test(boolean stripNativeCommands) throws Exception {
    JPackageCommand cmd = JPackageCommand.helloAppImage();
    cmd.addArguments("--mac-app-store", "--runtime-image", getRuntimeImage(stripNativeCommands));

    if (stripNativeCommands) {
        cmd.executeAndAssertHelloAppImageCreated();
    } else {
        cmd.execute(1);
    }
}
```

**Review by alexeysemenyukoracle — CHANGES_REQUESTED** [review]

File: test/jdk/tools/jpackage/macosx/MacAppStoreJlinkOptionsTest.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ * @build MacAppStoreJLinkOptionsTest
+ * @modules jdk.jpackage/jdk.jpackage.internal
+ * @requires (os.family == "mac")
+ * @run main/othervm -Xmx512m jdk.jpackage.test.Main
+ *  --jpt-run=MacAppStoreJLinkOptionsTest
+ */
+public class MacAppStoreJLinkOptionsTest {
+
+    @Test
+    public static void test() throws Exception {
```
Comment (alexeysemenyukoracle) [inline test/jdk/tools/jpackage/macosx/MacAppStoreJlinkOptionsTest.java:48]: I'd give some more descriptive names to test functions than `test` and `test2`. Something like `testWithStripNativeCommands` and `testWithoutStripNativeCommands` maybe?

File: test/jdk/tools/jpackage/macosx/MacAppStoreRuntimeTest.java (hunk 1 above)
Comment (sashamatveev) [reply test/jdk/tools/jpackage/macosx/MacAppStoreRuntimeTest.java]: Fixed.

File: test/jdk/tools/jpackage/macosx/MacAppStoreJlinkOptionsTest.java (hunk 2 above)
Comment (sashamatveev) [reply test/jdk/tools/jpackage/macosx/MacAppStoreJlinkOptionsTest.java]: Fixed.

**Comment by magicus** [comment]
We have the `NSMicrophoneUsageDescription` permission on the `java` launcher in the JDK, since otherwise no Java program can access the mike, even though most won't care. I agree that the situation is different for a jpackaged app, where the developer knows if that permission is needed or not.

Yes, plistbuddy is an official Apple program.

My understanding of the PR was that native commands are removed by jlink if the user is packaging on a mac for the App Store. I thought this was a workaround that solved the immediate problem of not being able to submit the app to App Store. (However, I don't know how the app is supposed to be started without a launcher...)

**Comment by alexeysemenyukoracle** [comment]
> However, I don't know how the app is supposed to be started without a launcher...
jpackage supplies alternative launcher that doesn't have plist.

**Review by kevinrushforth — APPROVED** [review]
I'm just catching up with this thread. Based on the analysis, it's pretty clear that any solution that actually allows bundling native launchers is going to take more research, and is out of scope for this bug. Some combination of [JDK-8286850](https://bugs.openjdk.java.net/browse/JDK-8286850) and/or providing Java launchers that don't have an `Info.plist` is likely needed to provide a complete solution.

Given that, I think that this PR seems a reasonable fix to avoid creating an app bundle that can't be uploaded to the app store.

**Pull request closed without merging**